File: engine-core/src/transposition.cpp
#include "gomoku/transposition.hpp"

namespace gomoku {

uint64_t TranspositionTable::next_power_of_2(uint64_t v) {
    if (v == 0) return 1;
    v--;
    v |= v >> 1;
    v |= v >> 2;
    v |= v >> 4;
    v |= v >> 8;
    v |= v >> 16;
    v |= v >> 32;
    return v + 1;
}

TranspositionTable::TranspositionTable(uint64_t size) {
    uint64_t actual = next_power_of_2(size);
    table_.resize(actual);
    mask_ = actual - 1;
}

std::optional<TTEntry> TranspositionTable::probe(uint64_t key) const {
    const auto& entry = table_[key & mask_];
    if (entry.key == key) {
        hits_++;
        return entry;
    }
    misses_++;
    return std::nullopt;
}

void TranspositionTable::store(uint64_t key, int score, int move, int depth, TTFlag flag) {
    auto& entry = table_[key & mask_];
    // Replacement policy: don't replace deeper entries for the same slot
    if (entry.key != 0 && entry.key != key && entry.depth > depth) {
        return;
    }
    entry.key   = key;
    entry.score = score;
    entry.move  = move;
    entry.depth = depth;
    entry.flag  = flag;
    stores_++;
}

std::optional<int> TranspositionTable::try_use_entry(uint64_t key, int depth, int alpha, int beta) {
    const auto& entry = table_[key & mask_];
    if (entry.key != key || entry.depth < depth) {
        return std::nullopt;
    }
    hits_++;

    if (entry.flag == TTFlag::Exact) {
        return entry.score;
    }
    if (entry.flag == TTFlag::LowerBound && entry.score >= beta) {
        return entry.score;
    }
    if (entry.flag == TTFlag::UpperBound && entry.score <= alpha) {
        return entry.score;
    }
    return std::nullopt;
}

int TranspositionTable::get_best_move(uint64_t key) const {
    const auto& entry = table_[key & mask_];
    if (entry.key == key) {
        return entry.move;
    }
    return -1;
}

void TranspositionTable::clear() {
    std::fill(table_.begin(), table_.end(), TTEntry{});
    hits_   = 0;
    misses_ = 0;
    stores_ = 0;
}

TTStats TranspositionTable::stats() const {
    TTStats s;
    s.hits   = hits_;
    s.misses = misses_;
    s.stores = stores_;
    // Count occupied entries
    size_t occupied = 0;
    for (const auto& e : table_) {
        if (e.key != 0) occupied++;
    }
    s.size = occupied;
    uint64_t total = hits_ + misses_;
    s.hit_rate = total > 0 ? (static_cast<double>(hits_) / total * 100.0) : 0.0;
    return s;
}

} // namespace gomoku

File: server/src/engine/transposition.mjs
// Transposition table for alpha-beta search
// Uses string hash keys (from GomokuBoard.hashKey()) in a JS Map

export const TT_EXACT = 0;
export const TT_LOWER = 1; // beta cutoff (score >= beta)
export const TT_UPPER = 2; // all-node (score <= alpha)

/**
 * @typedef {Object} TTEntry
 * @property {number} score
 * @property {number} move - best move found
 * @property {number} depth - search depth
 * @property {number} flag - TT_EXACT, TT_LOWER, or TT_UPPER
 */

export class TranspositionTable {
  /**
   * @param {number} maxEntries - maximum entries before eviction
   */
  constructor(maxEntries = 1 << 20) {
    this.maxEntries = maxEntries;
    this.table = new Map();
    this.hits = 0;
    this.misses = 0;
    this.stores = 0;
  }

  /**
   * Probe the table for a position.
   * @param {string} key - hash key from board.hashKey()
   * @returns {TTEntry|null}
   */
  probe(key) {
    const entry = this.table.get(key);
    if (entry) {
      this.hits++;
      return entry;
    }
    this.misses++;
    return null;
  }

  /**
   * Store a position in the table.
   * Replacement policy: always replace if new depth >= existing depth.
   */
  store(key, score, move, depth, flag) {
    const existing = this.table.get(key);
    if (existing && existing.depth > depth) {
      // Don't replace deeper entries
      return;
    }

    // Eviction if at capacity
    if (!existing && this.table.size >= this.maxEntries) {
      this._evict();
    }

    this.table.set(key, { score, move, depth, flag });
    this.stores++;
  }

  /**
   * Get the best move from TT for a position (for move ordering).
   */
  getBestMove(key) {
    const entry = this.table.get(key);
    return entry ? entry.move : -1;
  }

  /**
   * Try to get an exact score or usable bound from TT.
   * @returns {{ score: number, usable: boolean } | null}
   */
  tryUseEntry(key, depth, alpha, beta) {
    const entry = this.table.get(key);
    if (!entry || entry.depth < depth) return null;

    this.hits++;

    if (entry.flag === TT_EXACT) {
      return { score: entry.score, usable: true };
    }
    if (entry.flag === TT_LOWER && entry.score >= beta) {
      return { score: entry.score, usable: true };
    }
    if (entry.flag === TT_UPPER && entry.score <= alpha) {
      return { score: entry.score, usable: true };
    }

    return null;
  }

  /** Remove ~25% of entries when full */
  _evict() {
    const keys = this.table.keys();
    let count = (this.maxEntries * 0.25) | 0;
    for (const key of keys) {
      if (count-- <= 0) break;
      this.table.delete(key);
    }
  }

  clear() {
    this.table.clear();
    this.hits = 0;
    this.misses = 0;
    this.stores = 0;
  }

  get size() { return this.table.size; }

  stats() {
    const total = this.hits + this.misses;
    return {
      size: this.table.size,
      hits: this.hits,
      misses: this.misses,
      hitRate: total > 0 ? (this.hits / total * 100).toFixed(1) + '%' : 'N/A',
      stores: this.stores,
    };
  }
}
